api/src/main/java/javax/enterprise/context/spi/ContextController.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2016, Red Hat, Inc., and individual contributors
+ * by the @authors tag. See the copyright.txt in the distribution for a
+ * full listing of individual contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package javax.enterprise.context.spi;
+
+import javax.enterprise.context.ContextNotActiveException;
+import java.lang.annotation.Annotation;
+import java.util.Optional;
+
+/**
+ * A programmatic interface for retrieving, starting and stopping built in contexts
+ *
+ * The container provides a dependent scoped instance of this bean which may be injected to manage contexts
+ *
+ * <pre>
+ * &#064;Inject
+ * private ContextController contextController;
+ *
+ *
+ * ActiveContext context = contextController.getActiveContext(RequestScoped.class)
+ *                  .orElseGet(() -> contextController.activate(RequestScoped.class));
+ *
+ * // do some work.
+ *
+ * contextController.deactivate(context);
+ *
+ * </pre>
+ *
+ * @author John D. Ament
+ */
+public interface ContextController {
+
+    /**
+     * Obtains an active {@linkplain javax.enterprise.context.spi.ActiveContext context object} for the given
+     * {@linkplain javax.enterprise.context scope} based on what is active within the current thread.
+     *
+     * @param scopeType the {@linkplain javax.enterprise.context scope}
+     * @return an Optional representation of the {@linkplain javax.enterprise.context.spi.ActiveContext context object}
+     *         or else Optional.empty
+     * @throws IllegalArgumentException if there is more than one active context object for the given scope
+     */
+    Optional<ActiveContext> getActiveContext(Class<? extends Annotation> scopeType);
+
+    /**
+     * Obtains an active {@linkplain javax.enterprise.context.spi.ActiveContext context object} for the given
+     * {@linkplain javax.enterprise.context scope} based on its id.
+     * @param id the id of the context to be looked up
+     * @param scopeType the {@linkplain javax.enterprise.context scope}
+     * @return the {@linkplain javax.enterprise.context.spi.ActiveContext context object}
+     * @throws ContextNotActiveException if there is no active context object for the given scope with this ID
+     */
+    ActiveContext getActiveContext(String id, Class<? extends Annotation> scopeType);
```

Comment:
Neither of these "activated" contexts will be backed by an HTTP session so I don't think the ID makes any sense. It would only make sense if you allow to associate the context with the different storage (e.g. user-provided Map). See also [BoundConversationContext example](http://docs.jboss.org/weld/reference/latest/en-US/html/contexts.html#_managing_the_built_in_contexts). But this would complicate the API and I believe it's not worth including. I think we should either support a kind of **transient** conversations and sessions or don't support activation of session and conversation context at all.